### ISSUE TITLE
Ability to store gogs assets remote File system

### ISSUE THREAD

**Issue by abhilash07** [gogits/gogs#4362]
Hi everyone, it would be great if gogs extend the remote file system feature to store gogs assets (repositories, attachments, avatar & sessions). I mean remote storage is on s3, openstack swift, ceph storage and cloud foundry blob storage etc. Thanks

Regards

Abhilash S

**Comment by jhasse** [comment]
Wouldn't this work with GVfs / FUSE?

**Comment by abhilash07** [comment]
I didn't test, if you know please let us know how set up remote fs to store gogits assets?

**Comment by Unknwon** [comment]
What is the use case of this feature?

**Comment by abhilash07** [comment]
Good question, when we go for production such as cloud foundry computing so data size will be high? users want to store their data securily and safely. So that's the reason we are requesting to add feature to store in remote blob store(i.e., s3, azure or cloud foundry blobstore)

**Comment by Fastidious** [comment]
@abhilash07 but `git` isn't centralised. Anything online is already is one, and possibly, many computers. I really do not see the point. You can use FUSE to mount AWS S3 buckets on Linux, if that is what you are after, then configure `gogs` to save the repositories on it.

**Comment by Unknwon** [comment]
>Good question, when we go for production such as cloud foundry computing so data size will be high? users want to store their data securily and safely. So that's the reason we are requesting to add feature to store in remote blob store(i.e., s3, azure or cloud foundry blobstore)

The real huge thing is the repository, and Gogs currently couldn't help on that. Even if you store other things on so-called "cloud", you disk usage won't reduce much.

**Comment by lavvy** [comment]
@Fastidious how do you use an s3 mounted filesystem to store gogs volume? We have tried it but didn't work. 
Now we are using jgit to persist our repos to s3. @unknown you can as well give gogs an s3 backend with jgit, this could be quite easy to integrate to gogs. Please check this simple example http://www.fancybeans.com/blog/2012/08/24/how-to-use-s3-as-a-private-git-repository/

**Comment by Unknwon** [comment]
>Now we are using jgit to persist our repos to s3. @unknown you can as well give gogs an s3 backend with jgit, this could be quite easy to integrate to gogs.

Thanks for the info, but as far as jgit is a client tool, has no relation to server side software like Gogs.

**Comment by lavvy** [comment]
Jgit is as good as a distributed git. If one wants to use s3 backend, he
must install jgit in the server.
So the part you will do in gogs is to be able to key in the s3 or minio
credentials like bucket name, region, access key etc in your gogs settings
.

Then after a commit or push, gogs can then use the jgit installed on the
server along with the users credentials and push the repo to s3

**Comment by Unknwon** [comment]
I see.. but integrate with jgit is not a small and easy thing.

**Comment by lavvy** [comment]
I think gogs just need an ability to add remotes.
Then if a minio or s3 remote is specified then instead of git push, you
jgit push

I see.. but integrate with jgit is not a small and easy thing.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/gogits/gogs/issues/4362#issuecomment-292329169>, or mute
the thread

**Comment by Unknwon** [comment]
>Then if a minio or s3 remote is specified then instead of git push, you
jgit push

This sounds like a complete rewrite to Git backend to me...

**Comment by lavvy** [comment]
Yeah, and it coexists with git too. They can be interchangeable too.

**Comment by lavvy** [comment]
This is a really good feature. @unknown would you add the s3 support for
gogs?

**Comment by lavvy** [comment]
Any progress
